### ISSUE TITLE
Bob Test Failing Online

### ISSUE THREAD

**Issue by mick-io** [exercism/dart-test-runner#29]
The unit test are passing locally.  Getting the following error online:

```
An error occurred while running your tests. This might mean that there was an issue in our infrastructure,
or it might mean that you have something in your code that's causing our systems to break.

Please check your code, and if nothing seems to be wrong, try running the tests again.
```

`dart pub upgrade --precompile && dart run test --run-skipped` returns: `00:00 +25: All tests passed!`

`dart --version` returns `Dart SDK version: 2.13.4 (stable) (Wed Jun 23 13:08:41 2021 +0200) on "macos_x64"`

My solution follows:

```dart
class Bob {
  String response(String s) {
    s = s.trim();
    var hasLetters = RegExp(r"[A-Za-z]+").hasMatch(s);
    var isYelling = hasLetters && s == s.toUpperCase();
    var isQuestion = s.endsWith("?");

    if (s.isEmpty) return 'Fine. Be that way!';

    if (isYelling && isQuestion) return "Calm down, I know what I'm doing!";

    if (isYelling) return 'Whoa, chill out!';

    if (isQuestion) return "Sure.";

    return "Whatever.";
  }
}
```

**Comment by Stargator** [comment]
It may be related to this PR #28 , once the PR is merged I will ask you to try again

**Comment by Stargator** [comment]
#28 Issue seemed similar and has resolved itself, could you try again and let me know what happens?

**Comment by mick-io** [comment]
@Stargator I resubmitted the assignment just now to no avail.

**Comment by Stargator** [comment]
> @Stargator I resubmitted the assignment just now to no avail.

Ok, thanks

**Comment by Stargator** [comment]
@mick-io We recently update the test runner, please try again and let me know what happens.

**Comment by sevenseacat** [comment]
Another user (Ahmad Amin) has reported issues in Slack, running tests for the Two Fer example as well. 

https://exercism-team.slack.com/archives/CAQP7JL3T/p1631077238254900

**Comment by ErikSchierboom** [comment]
I've dug into this a little bit, and technically there is nothing wrong with the test runner. The only issue is that it is quite slow. I've 
tested with a two-fer solution (using `./bin/run-in-docker.sh two-fer /home/erik/exercism/dart/exercises/practice/two-fer /home/erik/exercism/dart/exercises/practice/two-fer`), and timed the two distinct dart-specific phases in `bin/run.sh`:

- dart pub upgrade: 1 second
- dart test: 18 seconds

Clearly, it's the actual testing that is slow and causing the website to error as the test runner runs into timeouts. Is there any way to speed up the testing?

**Comment by ErikSchierboom** [comment]
BTW I think this is duplicate of https://github.com/exercism/dart-test-runner/issues/31

**Comment by iHiD** [comment]
We should probably disable the test runner until we have got it performant, else we'll end up annoying the people using the Dart track. Any objections to that?

**Comment by Stargator** [comment]
> BTW I think this is duplicate of #31

@ErikSchierboom It's hard for me to say since they both get the same message (as far as I know), but I think the reasoning for the error message is not always the same.

> Clearly, it's the actual testing that is slow and causing the website to error as the test runner runs into timeouts. Is there any way to speed up the testing? @Stargator?

The documentation for `dart test` command is light. I can't even find documentation that the `--run-skipped` option exists.

Are you sure it's only running tests for the submitted solution? The only thing we can do is maybe adjust the time that is waited until we think there's a time out. Is there anything we can do on the website so it'll wait longer before calling it an error?

> We should probably disable the test runner until we have got it performant, else we'll end up annoying the people using the Dart track. Any objections to that?

@iHiD I would be fine with that, but I am having trouble getting the tests to time out when I run it locally. And so this is the only way I can test it.

But for the sake of users, yes let us disable it.

**Comment by Stargator** [comment]
@mick-io Are there any that are not affected? Or is this just a difference between any new exercises you work on and exercises you've already completed?

My meaning is, is this affecting you whenever you try to work a new exercise and it does not affect previous exercises you've completed? Or does this effect users even when they try the hello-world exercise?

I'm trying to determine what's the distinguishing factor. Maybe it's just for exercises that take longer than X seconds to complete?

**Comment by mick-io** [comment]
@Stargator  I'm not really sure because I noticed something curious today. My first iteration of Pangram passed the test locally but failed when I submitted it. However, after changing the code, my second iteration pass submission. 

### Iteration 1

```dart
/// Passes locally but fails when submitted.
class Pangram {
  final Runes alphaRunes = "abcdefghijklmnopqrstuvwxyz".runes;
  bool isPangram(String s) {
    var runes = alphaRunes.toList();
    for (var r in s.toLowerCase().runes) {
      var i = runes.indexOf(r);
      if (i != -1) runes[i] = -1;
    }
    return runes.every((r) => r == -1);
  }
}
```

### Iteration 2

```dart
/// Passes locally and when submitted.
class Pangram {
  final Runes alphaRunes = "abcdefghijklmnopqrstuvwxyz".runes;
  bool isPangram(String s) {
    var input = s.toLowerCase().runes;
    return alphaRunes.toList().where((r) => !input.contains(r)).isEmpty;
  }
}
```

**Comment by ErikSchierboom** [comment]
> The documentation for dart test command is light. I can't even find documentation that the --run-skipped option exists.

Yeah, I noticed the same thing. I tried `dart test --help` but did not immediately find anything that struck me as being potentially useful.

> Are you sure it's only running tests for the submitted solution? 

I'm sure it is only running the tests for the solution, yeah.

> The timeout could theoretically be increased, but it is set to the current limit to save on resources. I'll let @iHiD chime in.

The only thing we can do is maybe adjust the time that is waited until we think there's a time out. Is there anything we can do on the website so it'll wait longer before calling it an error?

> @iHiD I would be fine with that, but I am having trouble getting the tests to time out when I run it locally. And so this is the only way I can test it.

This is interesting. If you test an exercise via the Docker-based test runner, how long does that take on your machine?

**Comment by ErikSchierboom** [comment]
I've tried some more things, like running `dart test` on a dummy project in the `Dockerfile`, the idea being that maybe there were some things that are precompiled in the background, but that didn't help.
Sharding (see https://pub.dev/packages/test) also didn't help.
There is an open issue on slow tests: https://github.com/flutter/flutter/issues/69429

**Comment by Stargator** [comment]
@ErikSchierboom When I run the tests locally, they don't take 10s. Is there anything on the Docker side that could be optimized/cached?

@ErikSchierboom Can we try one build where there is either no time limit or double it to 20s? I'm thinking we need to know how long it currently takes to complete in order to get a better idea of the problem. Right now we just know it's incomplete at 10s. Will it complete at 11s? What if it won't complete until 2min? That could be very telling.

**Comment by ErikSchierboom** [comment]
@Stargator How fast is running the tests locally for you? Do you have Docker installed?

If I run the tests on the `example-success` solution in the test runner repo, it runs in:

```
real    0m13.392s
user    0m0.134s
sys     0m0.099s
```

I tested this by uncommenting the `docker build --rm -t exercism/test-runner .` in `./bin/run-in-docker.sh` and then running `time ./bin/run-in-docker.sh example-success $PWD/tests/example-success $PWD/tests/example-success`

> When I run the tests locally, they don't take 10s. Is there anything on the Docker side that could be optimized/cached?

I have no idea. I've tried lots of things, but have not managed to speed it up anymore. To me it looks like dart compilation is where the slowness comes from. 🤷 

> Can we try one build where there is either no time limit or double it to 20s? I'm thinking we need to know how long it currently takes to complete in order to get a better idea of the problem. Right now we just know it's incomplete at 10s. Will it complete at 11s? What if it won't complete until 2min? That could be very telling.

Not easily, but that's also not really necessary. We can run the Docker version of the test runner locally (as mentioned above). There will be _some_ difference in hardware, but that shouldn't make a huge difference.

**Comment by Sgoettschkes** [comment]
@ErikSchierboom I tested it locally on my machine using the new dart release (2.15) (see #40). Just this update changed the time to run from 20 sec to 11 sec on my old macbook pro. Maybe you could compare the two versions locally and see if the improvements are showing up on newer hardware as well and we could try running the test-runner in the production infrastructure again?

**Comment by ErikSchierboom** [comment]
Thanks for the heads-up! I've just checked this locally with the Docker-based setup (which is how we run the test runner in production), and the speedup is less pronounced, but still 2-3 seconds. 

@Stargator would upgrading to the `google/dart:2.15.1` docker image break anything?

**Comment by Sgoettschkes** [comment]
@ErikSchierboom Thanks for testing! The PR above is already merged, so the upgrade is already done if I understand the architecture correctly.

**Comment by ErikSchierboom** [comment]
@Sgoettschkes Ah, whoops! 

@mick-io Could you check to see if your Bob solution passes the test?

**Comment by Stargator** [comment]
> @Stargator would upgrading to the google/dart:2.15.1 docker image break anything?

I tested the track and no breakages were seen.

I approved the PR #40

Also, just fyi as @Sgoettschkes noted google/dart docker is deprecated in favor of dart, which was part of #40 that was reviewed and approved.

**Comment by Stargator** [comment]
@mick-io Have you been able to test the bob exercise to see if it still has the same issues?
